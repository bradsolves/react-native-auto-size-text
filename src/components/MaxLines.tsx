import * as React from 'react';
import {NativeSyntheticEvent, Text, TextLayoutEventData} from 'react-native';
import {AutoSizeTextProps} from '../types';

const MaxLines = (props: AutoSizeTextProps) => {
  const {fontSize, children, style, numberOfLines} = props;

  const [currentFont, setCurrentFont] = React.useState<number>(
    fontSize as number,
  );
  const handleTextMode = (e: NativeSyntheticEvent<TextLayoutEventData>) => {
    const {lines} = e.nativeEvent;
    if (lines.length > (numberOfLines as number)) {
      setCurrentFont(currentFont - 1);
    }
  };

  return (
    <Text
      numberOfLines={numberOfLines}
      adjustsFontSizeToFit
      style={[
        style,
        {
          fontSize: currentFont,
        },
      ]}
      onTextLayout={e => {
        handleTextMode(e);
      }}>
      {children}
    </Text>
  );
};

export default MaxLines;
