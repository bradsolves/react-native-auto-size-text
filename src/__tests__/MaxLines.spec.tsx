import { cleanup, render, waitFor } from '@testing-library/react-native';
import React from 'react';
import renderer from 'react-test-renderer';

import MaxLines from '../components/MaxLines';

afterEach(cleanup);

describe('Maxlines snapshot', () => {
  it('MaxLines renders correctly', () => {
    const tree = renderer
      .create(
        <MaxLines fontSize={15} numberOfLines={2}>
          Lorem impsum
        </MaxLines>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe('MaxLines mode', () => {
  it('should pass the right props to Text component inside MaxLines', async () => {
    const mockContent = 'Lorem ipsum';
    const mockNumberOfLines = 3;
    const mockFontSize = 15;
    const mockColor = 'red';

    const { getByTestId } = render(
      <MaxLines
        style={{ color: mockColor }}
        fontSize={mockFontSize}
        numberOfLines={mockNumberOfLines}
      >
        {mockContent}
      </MaxLines>
    );

    const textComponent = getByTestId('max-lines');
    const content = textComponent.props.children;
    const numberOfLines = textComponent.props.numberOfLines;
    const adjustsFontSizeToFit = textComponent.props.adjustsFontSizeToFit;
    const fontSize = textComponent.props.style[1].fontSize;
    const color = textComponent.props.style[0].color;

    await waitFor(() => {
      expect(content).toBe(mockContent);
      expect(numberOfLines).toBe(mockNumberOfLines);
      expect(fontSize).toBe(mockFontSize);
      expect(color).toBe(mockColor);
      expect(adjustsFontSizeToFit).toBeTruthy();
    });
  });
});
