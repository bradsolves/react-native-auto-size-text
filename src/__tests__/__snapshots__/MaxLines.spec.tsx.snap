// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Maxlines snapshot MaxLines renders correctly 1`] = `
<Text
  adjustsFontSizeToFit={true}
  numberOfLines={2}
  onTextLayout={[Function]}
  style={
    Array [
      undefined,
      Object {
        "fontSize": 15,
      },
    ]
  }
  testID="max-lines"
>
  Lorem impsum
</Text>
`;
